### PR TITLE
James 2694 restore annotations

### PR DESCRIPTION
**Pull request by remk**

### CODE REVIEW

File: mailbox/backup/src/main/java/org/apache/james/mailbox/backup/zip/ExtraFieldExtractor.java [hunk 1]
```diff
@@ -53,4 +67,29 @@
...
+        return getLongExtraField(SizeExtraField.ID_AJ, entry);
+    }
+
+    public static Optional<Date> getInternalDate(ZipEntry entry) throws ZipException {
+        ZipExtraField[] extraFields = ExtraFieldUtils.parse(entry.getExtra());
+        return Arrays.stream(extraFields)
+            .filter(field -> field.getHeaderId().equals(InternalDateExtraField.ID_AO))
+            .map(extraField -> ((InternalDateExtraField) extraField).getDateValue())
+            .findFirst()
+            .flatMap(Function.identity());
```
Comment (trantienduchn) [inline mailbox/backup/src/main/java/org/apache/james/mailbox/backup/zip/ExtraFieldExtractor.java]: I see the statemens:
```java
            .filter(field -> field.getHeaderId().equals(...))
            .map(extraField -> ((....) extraField).getDateValue())
            .findFirst()
            .flatMap(Function.identity());
```
duplicate many times, cany refactor it?

File: mailbox/backup/src/main/java/org/apache/james/mailbox/backup/zip/ZipEntryIterator.java [hunk 2]
```diff
@@ -77,4 +86,24 @@ public void close() throws IOException {
         zipInputStream.close();
     }
 
+    private Optional<ByteSource> getCurrentEntryContent(ZipEntry current) {
+        if (current.isDirectory()) {
+            return Optional.empty();
+        } else {
+            FileBackedOutputStream currentContent = new FileBackedOutputStream(fileThreshold, RESET_ON_FINALIZE);
```
Comment (trantienduchn) [inline mailbox/backup/src/main/java/org/apache/james/mailbox/backup/zip/ZipEntryIterator.java]: try with resource ?

File: mailbox/backup/src/main/java/org/apache/james/mailbox/backup/zip/MailAccountZipIterator.java [hunk 3]
```diff
@@ -91,17 +95,75 @@ private String getMailboxName(ZipEntry current) {
...
+            currentMailBox = Optional.empty();
+            return new MailboxWithAnnotationsArchiveEntry(getMailboxName(entry), getMailBoxId(entry).get(), NO_ANNOTATION);
+        } else {
+            // annotation directory => save current mailbox and fetch next entry
+            currentMailBox = Optional.of(new MailboxWithAnnotationsArchiveEntry(getMailboxName(entry), getMailBoxId(entry).get(), NO_ANNOTATION));
+            return next();
+        }
+    }
+
+    private boolean isLastEntryOrNextEntryIsNotOfType(Optional<ZipEntry> nextZipEntry, ZipEntryType mailboxAnnotationDir) throws ZipException {
```
Comment (trantienduchn) [inline mailbox/backup/src/main/java/org/apache/james/mailbox/backup/zip/MailAccountZipIterator.java]: can you rename `mailboxAnnotationDir`?

File: mailbox/backup/src/main/java/org/apache/james/mailbox/backup/zip/MailAccountZipIterator.java [hunk 4]
```diff
@@ -91,17 +95,75 @@ private String getMailboxName(ZipEntry current) {
...
+        return !nextZipEntry.isPresent() || !ExtraFieldExtractor.getEntryType(nextZipEntry.get()).equals(Optional.of(mailboxAnnotationDir));
```
Comment (trantienduchn) [inline mailbox/backup/src/main/java/org/apache/james/mailbox/backup/zip/MailAccountZipIterator.java]: using directly `Optional.of()` could leads to NullPointerException

File: mailbox/backup/src/main/java/org/apache/james/mailbox/backup/zip/MailAccountZipIterator.java [hunk 5]
```diff
@@ -91,17 +95,75 @@ private String getMailboxName(ZipEntry current) {
...
+        return isLastEntryOrNextEntryIsNotOfType(nextZipEntry, ZipEntryType.MAILBOX_ANNOTATION_DIR);
+    }
+
+    private boolean isLastEntryOrNextEntryIsNotOfType(Optional<ZipEntry> nextZipEntry) throws ZipException {
+        return isLastEntryOrNextEntryIsNotOfType(nextZipEntry, ZipEntryType.MAILBOX_ANNOTATION);
+    }
+
+    private MailArchiveEntry fromMailboxAnnotationEntry(ZipEntryWithContent current, Optional<ZipEntry> next) throws ZipException {
+        MailboxAnnotation annotation = MailboxAnnotation.newInstance(getAnnotationKey(currentMailBox.get(), current.getEntry()),
+            getAnnotationContent(current).orElse(""));
```
Comment (trantienduchn) [inline mailbox/backup/src/main/java/org/apache/james/mailbox/backup/zip/MailAccountZipIterator.java]: what is the meaning of "" ?

File: mailbox/backup/src/main/java/org/apache/james/mailbox/backup/zip/ZipEntryIterator.java (hunk 2 above)
Comment (remk) [reply mailbox/backup/src/main/java/org/apache/james/mailbox/backup/zip/ZipEntryIterator.java]: nice catch

File: mailbox/backup/src/main/java/org/apache/james/mailbox/backup/zip/ExtraFieldExtractor.java (hunk 1 above)
Comment (remk) [reply mailbox/backup/src/main/java/org/apache/james/mailbox/backup/zip/ExtraFieldExtractor.java]: nice catch

File: mailbox/backup/src/main/java/org/apache/james/mailbox/backup/zip/MailAccountZipIterator.java (hunk 4 above)
Comment (remk) [reply mailbox/backup/src/main/java/org/apache/james/mailbox/backup/zip/MailAccountZipIterator.java]: In this case the ZipEntryType is a value we create so we are confident it's not null. 
But your comment leads me to refactor this method. Thanks !

File: mailbox/backup/src/main/java/org/apache/james/mailbox/backup/SerializedMailboxId.java [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+    public String getValue() {
+        return value;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (o instanceof SerializedMailboxId) {
+            SerializedMailboxId that = (SerializedMailboxId) o;
+            return Objects.equal(value, that.value);
+        } else {
```
Comment (blackheaven) [inline mailbox/backup/src/main/java/org/apache/james/mailbox/backup/SerializedMailboxId.java]: You can drop it

File: mailbox/backup/src/main/java/org/apache/james/mailbox/backup/SerializedMessageId.java [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+    public String getValue() {
+        return value;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (o instanceof SerializedMessageId) {
+            SerializedMessageId that = (SerializedMessageId) o;
+            return Objects.equal(value, that.value);
+        } else {
```
Comment (blackheaven) [inline mailbox/backup/src/main/java/org/apache/james/mailbox/backup/SerializedMessageId.java]: You can drop it

File: mailbox/backup/src/main/java/org/apache/james/mailbox/backup/zip/MailAccountZipIterator.java [hunk 8]
```diff
@@ -108,8 +109,16 @@ private MailArchiveEntry fromMailboxEntry(ZipEntryWithContent current, Optional<
...
     }
 
-    private boolean isLastEntryOrNextEntryIsNotOfType(Optional<ZipEntry> nextZipEntry, ZipEntryType mailboxAnnotationDir) throws ZipException {
-        return !nextZipEntry.isPresent() || !ExtraFieldExtractor.getEntryType(nextZipEntry.get()).equals(Optional.of(mailboxAnnotationDir));
+    private boolean isOfType(ZipEntry nextZipEntry, ZipEntryType expectedType) {
+        Optional<ZipEntryType> zipEntryType = ExtraFieldExtractor.getEntryType(nextZipEntry);
+
+        return zipEntryType
+            .map(actualType -> actualType.equals(expectedType))
+            .orElseGet(() -> false);
```
Comment (blackheaven) [inline mailbox/backup/src/main/java/org/apache/james/mailbox/backup/zip/MailAccountZipIterator.java]: `orElse`

File: mailbox/backup/src/main/java/org/apache/james/mailbox/backup/zip/MailAccountZipIterator.java [hunk 9]
```diff
@@ -108,8 +109,16 @@ private MailArchiveEntry fromMailboxEntry(ZipEntryWithContent current, Optional<
         }
     }
 
-    private boolean isLastEntryOrNextEntryIsNotOfType(Optional<ZipEntry> nextZipEntry, ZipEntryType mailboxAnnotationDir) throws ZipException {
-        return !nextZipEntry.isPresent() || !ExtraFieldExtractor.getEntryType(nextZipEntry.get()).equals(Optional.of(mailboxAnnotationDir));
+    private boolean isOfType(ZipEntry nextZipEntry, ZipEntryType expectedType) {
+        Optional<ZipEntryType> zipEntryType = ExtraFieldExtractor.getEntryType(nextZipEntry);
+
+        return zipEntryType
+            .map(actualType -> actualType.equals(expectedType))
```
Comment (blackheaven) [inline mailbox/backup/src/main/java/org/apache/james/mailbox/backup/zip/MailAccountZipIterator.java]: `expectedType::equals`

File: mailbox/backup/src/main/java/org/apache/james/mailbox/backup/SerializedMailboxId.java (hunk 6 above)
Comment (remk) [reply mailbox/backup/src/main/java/org/apache/james/mailbox/backup/SerializedMailboxId.java]: Fixed in a previous  PR. Will be rebased

File: mailbox/backup/src/main/java/org/apache/james/mailbox/backup/SerializedMessageId.java (hunk 7 above)
Comment (remk) [reply mailbox/backup/src/main/java/org/apache/james/mailbox/backup/SerializedMessageId.java]: Fixed in a previous  PR. Will be rebased

**Comment by chibenwa** [comment]
This work had been merged as far as I can notice.

@remk can you confirm?

**Comment by chibenwa** [comment]
Many merge conflict, inactive for very long, opened on linagora repo and not the ASF one...

Reopen on ASF repository if relevant.

**Pull request closed without merging**